src/NodeAnalyzer/ModelAnalyzer.php

```diff
@@ -52,26 +73,122 @@ public function getPrimaryKey(string $class): ?string
     }
 
     /**
+     * @param  class-string<Model>|ObjectType  $model
+     */
+    public function isQueryScopeOnModel(string|ObjectType $model, string $scopeName, Scope $scope): bool
+    {
+        if (! is_string($model)) {
+            /** @var class-string<Model> $model */
+            $model = $model->getClassName();
+        }
+
+        $classReflection = $this->getClass($model);
+
+        if ($classReflection->hasMethod('scope' . ucfirst($scopeName))) {
+            return true;
+        }
+
+        try {
+            $method = $classReflection->getMethod($scopeName, $scope);
+
+            if ($this->usesScopeAttribute($method)) {
+                return true;
+            }
+        } catch (MissingMethodFromReflectionException) {
+        }
+
+        return false;
+    }
+
+    /**
+     * @param  class-string<Model>|ObjectType  $model
+     */
+    public function isRelationshipOnModel(string|ObjectType $model, string $relationName, Scope $scope): bool
+    {
+        if (! is_string($model)) {
+            /** @var class-string<Model> $model */
+            $model = $model->getClassName();
+        }
+
+        $classReflection = $this->getClass($model);
+
+        if (! $classReflection->hasMethod($relationName)) {
+            return false;
+        }
+
+        $extendedMethodReflection = $classReflection->getMethod($relationName, $scope);
+
+        foreach ($extendedMethodReflection->getVariants() as $extendedParametersAcceptor) {
+            if ($extendedParametersAcceptor->getReturnType()->isSuperTypeOf(self::relationType())->yes()) {
```

Comment:
```suggestion
            if (self::relationType()->isSuperTypeOf($extendedParametersAcceptor->getReturnType())->yes()) {
```

Reply:
Are you sure for this one? When I change it, the tests break for scenario `it_can_determine_if_the_model_has_relationship_method` at line `tests/NodeAnalyzer/ModelAnalyzerTest.php:117`

Reply:
Would PHPStan now infer the correct type purely from the fact it always returns a HasMany object anyways?

Reply:
Well you're right although I'm a little confused still because I assumed it would infer the type from the PHPDocs and the return of the method but I guess not.

Reply:
PHPStan doesn't descend into method and functions---it stops at the boundary. When inside a method phpstan checks that you return what you say you return, and when outside a method phpstan checks to make sure the arguments you pass are compatible 

https://phpstan.org/user-guide/troubleshooting-types#what-phpstan-doesn%E2%80%99t-do

Reply:
Ah okay. That makes sense. Thank you for your patience with me.

Reply:
No worries! It's a common misconception, but it helps to better understand how phpstan works once you grasp it :smiley: